drivers/i2s/i2s_cavs.c

```diff
@@ -312,52 +254,53 @@ static int i2s_cavs_configure(struct device *dev, enum i2s_dir dir,
 	u32_t frame_len = 0;
 	bool inverted_frame = false;
 
-	if (dir == I2S_DIR_TX) {
-		strm = &dev_data->tx;
-	} else {
-		SYS_LOG_ERR("TX direction must be selected");
-		return -EINVAL;
-	}
-
-	if (strm->state != I2S_STATE_NOT_READY &&
-	    strm->state != I2S_STATE_READY) {
-		SYS_LOG_ERR("invalid state");
+	if ((dev_data->tx.state != I2S_STATE_NOT_READY) &&
+			(dev_data->tx.state != I2S_STATE_READY) &&
+			(dev_data->rx.state != I2S_STATE_NOT_READY) &&
+			(dev_data->rx.state != I2S_STATE_READY)) {
+		SYS_LOG_ERR("invalid state tx(%u) rx(%u)", dev_data->tx.state,
+				dev_data->rx.state);
 		return -EINVAL;
 	}
 
 	if (i2s_cfg->frame_clk_freq == 0) {
-		strm->queue_drop(strm);
-		memset(&strm->cfg, 0, sizeof(struct i2s_config));
-		strm->state = I2S_STATE_NOT_READY;
-		return 0;
+		SYS_LOG_ERR("Invalid frame_clk_freq %u",
+				i2s_cfg->frame_clk_freq);
+		return -EINVAL;
 	}
 
 	if (word_size_bits < CAVS_SSP_WORD_SIZE_BITS_MIN ||
 	    word_size_bits > CAVS_SSP_WORD_SIZE_BITS_MAX) {
-		SYS_LOG_ERR("Unsupported I2S word size");
+		SYS_LOG_ERR("Unsupported I2S word size %u", word_size_bits);
 		return -EINVAL;
 	}
 
 	if (num_words < CAVS_SSP_WORD_PER_FRAME_MIN ||
 	    num_words > CAVS_SSP_WORD_PER_FRAME_MAX) {
-		SYS_LOG_ERR("Unsupported words per frame number");
+		SYS_LOG_ERR("Unsupported words per frame number %u", num_words);
 		return -EINVAL;
 	}
 
-	memcpy(&strm->cfg, i2s_cfg, sizeof(struct i2s_config));
+	if ((i2s_cfg->options & I2S_OPT_PINGPONG) == I2S_OPT_PINGPONG) {
```

Comment:
> I see that you have not supported ping-pong mode in this driver. In case of ping-pong mode for playback driver, the freeing of memory doesn't happen saving some valuable processing time (the same buffer keeps getting rewritten). Do you think this is not significant enough?

I decided to remove ping-pong mode because it is hard to track the buffers between the driver and application without proper interfaces defined in the APIs. Since the DMA will be free-running, if the application is "slow", for example, the DMA might have moved on and the application will not know which buffer the DMA is actively using.